x/programs/rust/sdk_macros/src/lib.rs

```diff
@@ -24,50 +16,127 @@ fn convert_param(param_name: &Ident) -> proc_macro2::TokenStream {
-        // A typed argument is a parameter. An untyped (receiver) argument is a `self` parameter.
-        if let FnArg::Typed(PatType { pat, ty, .. }) = fn_arg {
-            // ensure first parameter type is Program
-            if index == 0 && !is_program(ty) {
-                panic!("First parameter must be Program.");
-            }
-            if let Pat::Ident(ref pat_ident) = **pat {
-                return (&pat_ident.ident, quote! { i64 });
-            }
-            // add unused variable
-            if let Pat::Wild(_) = **pat {
-                if is_program(ty) {
-                    return (&empty_param, quote! { i64 });
-                } else {
-                    panic!("Unused variables only supported for Program.")
+    // TODO:
+    // prefix with an underscore
+    let new_name = Ident::new(&format!("{name}_guest"), name.span());
+
+    // to be used as the result below
+    let first_arg_err = match input_args.first() {
+        Some(FnArg::Typed(PatType { ty, .. })) if is_program(ty) => None,
+        arg => {
+            let err = match arg {
+                Some(FnArg::Typed(PatType { ty, .. })) => {
+                    syn::Error::new(
+                        ty.span(),
+                        format!("The first paramter of a function with the `#[public]` attribute must be of type `{PROGRAM_TYPE}`"),
+                    )
+                }
+                Some(_) => {
+                    syn::Error::new(
+                        arg.span(),
+                        format!("The first paramter of a function with the `#[public]` attribute must be of type `{PROGRAM_TYPE}`"),
+                    )
                 }
+                None => {
+                    syn::Error::new(
+                        input.sig.paren_token.span.join(),
+                        format!("Functions with the `#[public]` attribute must have at least one parameter and the first parameter must be of type `{PROGRAM_TYPE}`"),
+                    )
+                }
+            };
+
+            Some(err)
+        }
+    };
+
+    let param_idents = input_args
+        .iter()
+        .enumerate()
+        .skip(1)
+        .map(|(i, fn_arg)| match fn_arg {
+            FnArg::Receiver(_) => Err(syn::Error::new(
+                fn_arg.span(),
+                "Functions with the `#[public]` attribute cannot have a `self` parameter.",
+            )),
+            FnArg::Typed(PatType { pat, .. }) => match pat.as_ref() {
+                // TODO:
+                // we should likely remove this constraint. If we provide upgradability
+                // in the future, we may not want to change the function signature
+                // which means we might want wildcards in order to help produce stable APIs
+                Pat::Wild(_) => Err(syn::Error::new(
+                    fn_arg.span(),
+                    "Functions with the `#[public]` attribute can only ignore the first parameter.",
```

Comment:
let's just remove it then? Why TODO for later?

Reply:
Didn't want to just make the executive decision. The TODO is me punting the discussion.

You still think I should just get rid of it? It's easier to delete later than to re-add, that's why I added the TODO (though re-adding is not hard).